runner/runner.go

```diff
@@ -0,0 +1,379 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"math/rand"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"sync/atomic"
+	"syscall"
+	"time"
+)
+
+type siteInfo struct {
+	Multisite int
+	Siteurl   string
+	Disabled  int
+}
+
+type site struct {
+	URL string
+}
+
+type event struct {
+	URL       string
+	Timestamp int
+	Action    string
+	Instance  string
+}
+
+var (
+	wpCliPath string
+	wpNetwork int
+	wpPath    string
+
+	workersGetEvents int
+	workersRunEvents int
+
+	heartbeatInt int
+
+	disabledLoopCount    uint64
+	eventRunErrCount     uint64
+	eventRunSuccessCount uint64
+
+	logger  *log.Logger
+	logDest string
+	debug   bool
+)
+
+const getEventsLoop time.Duration = time.Minute
+const getEventsBreak time.Duration = time.Second
+const runEventsBreak time.Duration = time.Second * 10
+
+func init() {
+	flag.StringVar(&wpCliPath, "cli", "/usr/local/bin/wp", "Path to WP-CLI binary")
+	flag.IntVar(&wpNetwork, "network", 0, "WordPress network ID, `0` to disable")
+	flag.StringVar(&wpPath, "wp", "/var/www/html", "Path to WordPress installation")
+	flag.IntVar(&workersGetEvents, "workers-get", 1, "Number of workers to retrieve events")
+	flag.IntVar(&workersRunEvents, "workers-run", 5, "Number of workers to run events")
+	flag.IntVar(&heartbeatInt, "heartbeat", 60, "Heartbeat interval in seconds")
+	flag.StringVar(&logDest, "log", "os.Stdout", "Log path, omit to log to Stdout")
+	flag.BoolVar(&debug, "debug", false, "Include additional log data for debugging")
+	flag.Parse()
+
+	setUpLogger()
+
+	// TODO: Should check for wp-config.php instead?
+	validatePath(&wpCliPath)
+	validatePath(&wpPath)
+}
+
+func main() {
+	logger.Printf("Starting with %d event-retreival worker(s) and %d event worker(s)", workersGetEvents, workersRunEvents)
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
+
+	sites := make(chan site)
+	events := make(chan event)
+
+	go spawnEventRetrievers(sites, events)
+	go spawnEventWorkers(events)
+	go retrieveSitesPeriodically(sites)
+	go heartbeat()
+
+	caughtSig := <-sig
+	logger.Printf("Stopping, got signal %s", caughtSig)
+}
+
+func spawnEventRetrievers(sites <-chan site, queue chan<- event) {
+	for w := 1; w <= workersGetEvents; w++ {
+		go queueSiteEvents(w, sites, queue)
+	}
+}
+
+func spawnEventWorkers(queue <-chan event) {
+	workerEvents := make(chan event)
+
+	for w := 1; w <= workersRunEvents; w++ {
+		go runEvents(w, workerEvents)
+	}
+
+	for event := range queue {
+		workerEvents <- event
+	}
+
+	close(workerEvents)
+}
+
+func retrieveSitesPeriodically(sites chan<- site) {
+	for {
```

Comment:
Instead of calling `time.Sleep` between runs, we can do:

```
for range time.Tick(time.Second) {
}
```

You can convert `getEventsLoops` to a `time.Duration` and pass that to `time.Tick()`.